### ISSUE TITLE
CtTypeReference: Inconsistent API documentation

### ISSUE THREAD

**Issue by msteinbeck** [INRIA/spoon#1752]
According to `CtTypeReference`, `getActualClass` returns null if the requested class is not found and throws a `SpoonClassNotFoundException`. Which one is correct?

```java
/**
 * Gets the Java runtime class of the referenced type.
 *
 * This is a low-level feature, it should never been used.
 * Use {@link #getTypeDeclaration()} instead,
 * in order to only stay in the Spoon world and manipulate CtType instead of java.lang.Class
 *
 * @return the Java class or null if the class is not found (not in
 * classpath)
 * @throws SpoonClassNotFoundException
 * 		if the class is not in the classpath
 */
Class<T> getActualClass();
```

**Comment by surli** [comment]
From what I can see in the implementation, you will get a `SpoonClassNotFoundException` if the class is indeed not in the classpath.
However in some rare cases, you can get a `null`, when for example you try to call the method on a `CtTypeParameter` with no bounding class.

@monperrus @pvojtechovsky WDYT? Should we just improve the documentation, or throws SpoonClassNotFound also in the few cases where we currently returns null?

**Comment by msteinbeck** [comment]
I found this because I have an issue in no classpath mode. I have a reference to a class not being available in my classpath. When trying to resolve this reference using `getDeclaration`, Spoon throws a `SpoonClassNotFoundException` even though I enabled noclasspath mode explicitly. This is not true for every reference, though. I'm currently working on a minimal example emphasizing this issue.

**Comment by surli** [comment]
> When trying to resolve this reference using getDeclaration, Spoon throws a SpoonClassNotFoundException even though I enabled noclasspath mode explicitly.

This behaviour is expected: the no classpath mode is used to build the Spoon model, even if you don't have the full classpath. However, here you're trying to get a class that is not reachable: so, it throws an exception, whatever the mode is. 
You have to catch and treat those exceptions yourself here.

**Comment by msteinbeck** [comment]
But `getDeclaration` says that is returns `null` if the declaration is not available in classpath:

```java
/**
 * Returns the {@link CtType}, that corresponds to the
 * reference or <code>null</code> if the type declaration is not in the
 * analyzed source files,
 *
 * {@link #getTypeDeclaration()} is a newer and better alternative that never returns null.
 *
 * @return the referenced element or <code>null</code> if the type
 * declaration is not the analyzed source files.
 */
@DerivedProperty
CtType<T> getDeclaration();
```

**Comment by surli** [comment]
My bad I thought you were using `getActualClass`.
Then yes, it looks like a bug in `getDeclaration`: could you provide a unit test in Spoon to display where the bug happens exactly?

**Comment by msteinbeck** [comment]
As I said, providing a test case isn't that simple, but I gave you the lines that fail: https://github.com/INRIA/spoon/issues/1753

**Comment by surli** [comment]
Hi @msteinbeck 

so I proposed #1761 to clarify the documentation of getActualClass and #1762 to fix the bug presented in #1753. 

> When trying to resolve this reference using getDeclaration, Spoon throws a SpoonClassNotFoundException even though I enabled noclasspath mode explicitly. This is not true for every reference, though. I'm currently working on a minimal example emphasizing this issue

Concerning this issue, I did not manage to reproduce it: does it concern a specific type of reference? Any more hints about it?

**Comment by msteinbeck** [comment]
> Concerning this issue, I did not manage to reproduce it: does it concern a specific type of reference? Any more hints about it?
Let me try to find a minimal example for pr #1753. Can I create a pr for a pr?

**Comment by surli** [comment]
> Can I create a pr for a pr?

You can create a PR in my own fork to the branch of my PR :)

**Comment by msteinbeck** [comment]
You should now have a pr for your pr.

**Comment by surli** [comment]
Yes, thanks! It looks like a tricky one :)

**Comment by surli** [comment]
Closed by #1761
